src/insertItem.ts

```diff
@@ -0,0 +1,449 @@
+        }
+        return await this.insertInObject(template, textEditor, templateKeys.parameters, parameter, name);
+    }
+
+    // tslint:disable-next-line:no-any
+    private async insertInObject(template: DeploymentTemplate, textEditor: vscode.TextEditor, part: string, data: any, name: string): Promise<boolean> {
+        let templatePart = this.getTemplateObjectPart(template, part);
+        if (!templatePart) {
+            let topLevel = template.topLevelValue;
+            if (!topLevel) {
+                vscode.window.showErrorMessage('Invalid ARM template!');
+                return false;
+            }
+            // tslint:disable-next-line:no-any
+            let subPart: any = {};
+            // tslint:disable-next-line:no-unsafe-any
+            subPart[name] = data;
+            await this.insertInObjectInternal(topLevel, textEditor, subPart, part, 1);
+        } else {
+            await this.insertInObjectInternal(templatePart, textEditor, data, name);
+        }
+        return true;
+    }
+
+    // tslint:disable-next-line:no-any
+    private async insertInObjectInternal(templatePart: Json.ObjectValue, textEditor: vscode.TextEditor, data: any, name: string, indentLevel: number = 2): Promise<number> {
+        let isFirstItem = templatePart.properties.length === 0;
+        let startText = isFirstItem ? '' : ',';
+        let index = isFirstItem ? templatePart.span.endIndex : templatePart.properties[templatePart.properties.length - 1].span.afterEndIndex;
+        let tabs = '\t'.repeat(indentLevel - 1);
+        let endText = isFirstItem ? `\r\n${tabs}` : ``;
+        let text = typeof (data) === 'object' ? JSON.stringify(data, null, '\t') : `"${data}"`;
+        let indentedText = this.indent(`\r\n"${name}": ${text}`, indentLevel);
+        return await this.insertText(textEditor, index, `${startText}${indentedText}${endText}`);
+    }
+
+    private async insertVariable(template: DeploymentTemplate, textEditor: vscode.TextEditor): Promise<boolean> {
+        let name = await this.ui.showInputBox({ prompt: "Name of variable?" });
+        return await this.insertInObject(template, textEditor, templateKeys.variables, insertCursorText, name);
+    }
+
+    private async insertOutput(template: DeploymentTemplate, textEditor: vscode.TextEditor): Promise<boolean> {
+        let name = await this.ui.showInputBox({ prompt: "Name of output?" });
+        const outputType = await this.ui.showQuickPick(getItemType(), { placeHolder: 'Type of output?' });
+        let output: Output = {
+            type: outputType.value,
+            value: insertCursorText.replace(/"/g, '')
+        };
+        return await this.insertInObject(template, textEditor, templateKeys.outputs, output, name);
+    }
+    private async insertFunctionAsTopLevel(topLevel: Json.ObjectValue | undefined, textEditor: vscode.TextEditor): Promise<boolean> {
+        if (!topLevel) {
+            vscode.window.showErrorMessage('Invalid ARM template!');
```

Comment:
Instead of showErrorMessage, please throw.  Our registerCommand code will catch it and automatically send error telemetry and show a message.  Same for others in this file.

If you pass through the actionContext parameter from the registerCommand, you can set `actionContext.errorHandling.suppressReportIssue = true` before throwing to keep the "report an issue" button from showing up in the error dialog.

Reply:
Then insertInObject/etc probably don't need to return bool

Reply:
Fixed